### DISCUSSION TITLE
Ragel for protocols

### DISCUSSION THREAD

**Discussion by burdiyan** [adrian-thurston/ragel#95] [Q&A]
I remember reading a blog post by Zed Shaw who was talking about how he was using Ragel for protocols and flow control. I.e. using Ragel not as a parser, but as a general-purpose state machine.

The post itself is available through the Internet Archive: https://web.archive.org/web/20070116050758/http://www.zedshaw.com/tips/ragel_state_charts.html

I guess this is not a very common approach for using Ragel, but I kinda like it. I wonder if more people have experience with using Ragel in such a way.

**Comment by remcycles** [comment]
Yes, I am also interested in using Ragel for compiling state machines as well.  Thanks for the link.  That article was an interesting read and provided good examples of Ragel state charts.  I didn't quite understand them from the docs (no large examples there).

I previously wrote about using regular expressions in Ragel to implement line coding for a stream of binary data and generate a WaveDrom diagram.  That was a fun practice project, but it would be nice to try some state charts too.
https://remcycles.net/blog/ragel_line_code.html

A little off-topic, but I think the Quantum Leaps tools are a better fit for my typical use cases (embedded systems development). https://www.state-machine.com/
